File: rust/src/medium/add_two_numbers.rs
// https://leetcode.com/problems/add-two-numbers/description/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub fn add_two_numbers(
    l1: Option<Box<ListNode>>,
    l2: Option<Box<ListNode>>,
) -> Option<Box<ListNode>> {
    let mut additional = None;

    let mut l1_pointer = l1;
    let mut l2_pointer = l2;

    let mut v = Vec::new();

    let mut result = None;

    while l1_pointer.is_some() || l2_pointer.is_some() {
        let mut sum = additional.unwrap_or(0);
        if let Some(l1_node) = l1_pointer {
            sum += l1_node.val;
            l1_pointer = l1_node.next;
        }
        if let Some(l2_node) = l2_pointer {
            sum += l2_node.val;
            l2_pointer = l2_node.next;
        }

        if sum >= 10 {
            additional = Some(sum / 10);
            sum %= 10;
        } else {
            additional = None;
        }

        v.push(sum);
    }

    if let Some(val) = additional {
        v.push(val);
    }

    for item in v.iter().rev() {
        let list_node = ListNode {
            val: *item,
            next: result,
        };

        result = Some(Box::new(list_node));
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        // `$ cargo test -- add_two_numbers --nocapture`

        let l1_1 = ListNode::new(2);

        let mut l1_2 = ListNode::new(3);
        l1_2.next = Some(Box::new(l1_1));

        let l1_3 = ListNode {
            val: 4,
            next: Some(Box::new(l1_2)),
        };

        let l2_1 = ListNode::new(9);

        let mut l2_2 = ListNode::new(3);
        l2_2.next = Some(Box::new(l2_1));

        let l2_3 = ListNode {
            val: 8,
            next: Some(Box::new(l2_2)),
        };

        // 234 + 938 = 1172
        println!(
            "{:?}",
            add_two_numbers(Some(Box::new(l1_3.clone())), Some(Box::new(l2_3.clone())))
        );

        let l2_4 = ListNode {
            val: 1,
            next: Some(Box::new(l2_3)),
        };

        // 234 + 9381 = 9615
        println!(
            "{:?}",
            add_two_numbers(Some(Box::new(l1_3)), Some(Box::new(l2_4)))
        );
    }
}

File: cpp/medium/add_two_numbers.cpp
// https://leetcode.com/problems/add-two-numbers/

#include <iostream>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    static ListNode* addTwoNumbers(ListNode* l1, ListNode* l2);
};

void pushValBack(ListNode **head, ListNode **end, int val) {
    ListNode *l3 = new ListNode(val);
    if (*head == nullptr) {
        *head = l3;
        *end = *head;
    } else {
        (*end)->next = l3;
        *end = (*end)->next;
    }
}

ListNode* Solution::addTwoNumbers(ListNode* l1, ListNode* l2)
{
    int additional = 0;
    int sum = 0;
    ListNode *head = nullptr;
    ListNode *end = nullptr;

    while (l1 != nullptr && l2 != nullptr) {
        sum = l1->val + l2->val + additional;
        if (sum >= 10) {
            additional = 1;
            sum -= 10;
        } else {
            additional = 0;
        }

        pushValBack(&head, &end, sum);

        l1 = l1->next;
        l2 = l2->next;
    }

    while (l1 != nullptr) {
        sum = l1->val + additional;
        if (sum >= 10) {
            additional = 1;
            sum -= 10;
        } else {
            additional = 0;
        }

        pushValBack(&head, &end, sum);

        l1 = l1->next;
    }

    while (l2 != nullptr) {
        sum = l2->val + additional;
        if (sum >= 10) {
            additional = 1;
            sum -= 10;
        } else {
            additional = 0;
        }

        pushValBack(&head, &end, sum);

        l2 = l2->next;
    }

    if (additional != 0) {
        pushValBack(&head, &end, additional);
    }

    return head;
}

int main() {
    auto l1_1 = ListNode(2);
    auto l1_2 = ListNode(3, &l1_1);
    auto l1_3 = ListNode(4, &l1_2);

    auto l2_1 = ListNode(5);
    auto l2_2 = ListNode(6, &l2_1);
    auto l2_3 = ListNode(7, &l2_2);
    auto l2_4 = ListNode(7, &l2_3);

    // 234 + 567 = 801 
    auto l3 = Solution::addTwoNumbers(&l2_3, &l1_3);
    auto l4 = l3;

    while (l4 != nullptr){
        std::cout << l4->val;
        l4 = l4->next;
        delete l3;
        l3 = l4;
    }

    std::cout << std::endl;

    // 234 + 5677 = 5911 
    l3 = Solution::addTwoNumbers(&l2_4, &l1_3);
    l4 = l3;

    while (l4 != nullptr){
        std::cout << l4->val;
        l4 = l4->next;
        delete l3;
        l3 = l4;
    }

    std::cout << std::endl;

    return 0;
}